src/device/mod.rs

```diff
@@ -253,42 +260,38 @@ impl<T: Tun, S: Sock> DeviceHandle<T, S> {
         }
     }
 
-    fn event_loop(_i: usize, device: &Lock<Device<T, S>>) {
-        #[cfg(target_os = "linux")]
-        let mut thread_local = ThreadData {
-            src_buf: [0u8; MAX_UDP_SIZE],
-            dst_buf: [0u8; MAX_UDP_SIZE],
-            iface: if _i == 0 || !device.read().config.use_multi_queue {
-                // For the first thread use the original iface
-                Arc::clone(&device.read().iface)
-            } else {
-                // For for the rest create a new iface queue
-                let iface_local = Arc::new(
-                    T::new(&device.read().iface.name().unwrap())
-                        .unwrap()
-                        .set_non_blocking()
-                        .unwrap(),
-                );
-
-                device
-                    .read()
-                    .register_iface_handler(Arc::clone(&iface_local))
-                    .ok();
-
-                iface_local
-            },
-        };
+    #[cfg(not(target_os = "linux"))]
+    pub fn set_iface(&mut self, new_iface: T) -> Result<(), Error> {
+        // Even though device struct is not being written to, we still take a write lock on device to stop the event loop
+        self
+            .device
+            .read()
+            .try_writeable(
+                |device| device.trigger_yield(),
+                |device| {
+                    (device.update_seq, _) = device.update_seq.overflowing_add(1);
+                    unsafe {
+                        device.queue.clear_event_by_fd(device.iface.as_raw_fd());
+                    }
+                    device.iface = Arc::new(new_iface.set_non_blocking()?);
+                    device.register_iface_handler(device.iface.clone())?;
+                    device.cancel_yield();
 
-        #[cfg(not(target_os = "linux"))]
-        let mut thread_local = ThreadData {
-            src_buf: [0u8; MAX_UDP_SIZE],
-            dst_buf: [0u8; MAX_UDP_SIZE],
-            iface: Arc::clone(&device.read().iface),
-        };
+                    Ok(())
+                }
+            ).ok_or(Error::SetTunnel("Failed to get device lock when setting tunnel".to_string()))?
```

Comment:
I see that this is the only possible error case, leave Error as plain enum value, add this message to rustdic of that value.